lib/ucfg.rb

```diff
@@ -43,20 +43,27 @@ def self.validate_recursively(config, schema, config_path)
     # fails if property is provided as other type
     config.each do |key, value|
       if schema["properties"].key?(key) && schema["properties"][key].key?("type")
-        if schema["properties"][key]["type"] != value_type(value)
+        if schema["properties"][key]["type"].include?(value_type(value))
+         valid = true
+        elsif value.nil?
+          valid = false
+          errors << "Property `#{(config_path + [key]).join('.')}` must be of type `#{(schema['properties'][key]['type']).first}` or `#{(schema['properties'][key]['type']).last}` (provided `#{value_type(value)}`)"
+        else
           valid = false
           errors << "Property `#{(config_path + [key]).join('.')}` must be of type `#{schema['properties'][key]['type']}` (provided value `#{value}` of type `#{value_type(value)}`)"
         end
       end
     end
 
-    schema["properties"].each do |key, value|
-      if config[key].is_a?(Hash)
-        new_config = config[key]
-        new_schema = value
-        recursion_result = validate_recursively(new_config, new_schema, config_path + [key])
-        errors.concat(recursion_result.errors)
-        valid = false unless recursion_result.valid?
+    if(!schema["properties"].nil?)
```

Comment:
The parenthesis are not needed in Ruby, and I think something like the following would be a bit cleaner:

```suggestion
    if schema.key?("properties")
```

Reply:
Hmm yes I complicated with this...Changed